src/crypto/OutgoingRoomKeyRequestManager.js

```diff
@@ -107,6 +110,79 @@ export default class OutgoingRoomKeyRequestManager {
         });
     }
 
+    /**
+     * Cancel room key requests, if any match the given details
+     *
+     * @param {module:crypto~RoomKeyRequestBody} requestBody
+     *
+     * @returns {Promise} resolves when the request has been updated in our
+     *    pending list.
+     */
+    cancelRoomKeyRequest(requestBody) {
+        return this._cryptoStore.getOutgoingRoomKeyRequest(
+            requestBody,
+        ).then((req) => {
+            if (!req) {
+                // no request was made for this key
+                return;
+            }
+
+            if (req.state === ROOM_KEY_REQUEST_STATES.UNSENT) {
+                // just delete it
+                console.log(
+                    'deleting unnecessary room key request for ' +
+                    stringifyRequestBody(requestBody),
+                );
+                return this._cryptoStore.deleteOutgoingRoomKeyRequest(
+                    req.requestId, ROOM_KEY_REQUEST_STATES.UNSENT,
+                );
+            }
+
+            if (req.state !== ROOM_KEY_REQUEST_STATES.SENT) {
+                // the only other option is cancellation_pending, in which case
+                // there is nothing else to do here
+            }
```

Comment:
> I feel you may be missing something here?

apparently! thank you!

> As an aside: I dislike and would discourage the use of "handle 2 cases, only 1 left so don't bother to check" since this will break in weird ways if we end up adding more states

I don't really understand what you want me to do here. If I explicitly handle each case, it's *still* going to break in weird ways if we end up adding more states. I've rearranged the code a bit. I think it's clear, but you may disagree.

I've also added a FIXME because I've spotted another edge case :-( :-(